DOCS-906:  Document `wheeled-odometry` model of movement sensor (#1734)
viamrobotics/docs: docs/components/movement-sensor/wheeledodometry.md

---
title: "Configure a wheeledodometry movement sensor"
linkTitle: "wheeledodometry"
type: "docs"
description: "Configure a wheeledodometry movement sensor."
images: ["/icons/components/imu.svg"]
tags: ["movement sensor", "components", "movement sensor"]
# SMEs: Rand, Martha
---

Configure a `wheeledodometry` movement sensor to implement _wheeled odometry_ on your robot.

_Wheeled odometry_ is the estimation of position, orientation, linear velocity, and angular velocity using the dimensions of a base.
This model uses [encoders](/components/encoder/) to get an odometry estimate from an encoder wheeled base.

## Set-up requirements

To prepare your robot, attach [encoders](/components/encoder/) to each the motors of your base to measure their rotation.

> - Configure each of these encoded motors [as encoder components](/components/encoder/#configuration).

Author: ```suggestion
- Configure your rover as a [wheeled base component](/components/base/wheeled/).
- Configure each of these encoded motors [as encoder components](/components/encoder/#configuration).
```

"name" : "<your-wheeledodometry-movement-sensor-name>",
"type" : "movement_sensor",
"model" : "wheeledodometry",
"attributes" : {
    "base" : "<your-base-name>",
    "left_motors" : ["<your-base-left-motor-name-1>", "<your-base-left-motor-name-2>"],
    "right_motors" : ["<your-base-right-motor-name-1", "your-base-right-motor-name-2>"],
    "time-interval-msec": <number>
}
```

Fill in and edit the attributes as applicable.

## Attributes

The following attributes are available for `wheeledodometry` movement sensors:

| Name | Type | Inclusion | Description |
| ---- | ---- | --------- | ----------- |
| `base` | string | **Required** | The `name` of the [base](/components/base/) to which the encoders making up this movement sensor are wired. |
| `left_motors` | object | **Required** | A struct holding the name of each of the bases' left [encoded motors](/components/encoder/). |
| `right_motors` | object | **Required** | A struct holding the name of each of the bases' right [encoded motors](/components/encoder/). |
| `time_interval_msec` | number | Optional | The time in between each wheeled odometry calculation. <br> Default: `500.0` </br> |

> With a configured `wheeledodometry` movement sensor, after every time `time_interval_msec` elapses during a [session](/program/apis/sessions), your robot calculates an estimation of the position, orientation, linear velocity, and angular velocity of the wheeled base.

Author: ```suggestion
With a configured `wheeledodometry` movement sensor, after every time `time_interval_msec` elapses during a [session](/program/apis/sessions/), your robot calculates an estimation of the position, orientation, linear velocity, and angular velocity of the wheeled base.
```

---
title: "Configure a wheeledodometry movement sensor"

> linkTitle: "wheeledodometry"

Reviewer: ```suggestion
linkTitle: "wheeled-odometry"
```

Author: Done

type: "docs"

> description: "Configure a wheeledodometry movement sensor."

Reviewer: ```suggestion
description: "Configure a wheeled-odometry movement sensor."
```

Author: Done

images: ["/icons/components/imu.svg"]
tags: ["movement sensor", "components", "movement sensor"]
# SMEs: Rand, Martha
---

> Configure a `wheeledodometry` movement sensor to implement _wheeled odometry_ on your robot.

Reviewer: ```suggestion
Configure a `wheeled-odometry` movement sensor to implement _wheeled odometry_ on your robot.
```

Author: Done

_Wheeled odometry_ is the estimation of position, orientation, linear velocity, and angular velocity using the dimensions of a base.

> This model uses [encoders](/components/encoder/) to get an odometry estimate from an encoder wheeled base.

Reviewer: ```suggestion
This model uses [encoders](/components/encoder/) from Position Reporting motors to get an odometry estimate from a wheeled base.
```

Author: Done

## Set-up requirements

> To prepare your robot, attach [encoders](/components/encoder/) to each the motors of your base to measure their rotation.

Reviewer: Bit more explanation on required hardware.

- Configure your rover as a [wheeled base component](/components/base/wheeled/).

> - Configure each of these encoded motors [as encoder components](/components/encoder/#configuration).

Reviewer: these are motor components. They have to be position reporting.

Author: Oh, ok! Will update to reflect and comment here when done. What kinds of motors specifically will work? Is it just https://docs.viam.com/components/motor/gpio/ or am I confused?

Author: Think I updated but happy to change all around again if it's not just GPIO or isn't GPIO-- apologies for the confusion!

## Configuration

Navigate to the **Config** tab of your robot's page in [the Viam app](https://app.viam.com).
Click on the **Components** subtab and navigate to the **Create component** menu.
Select **Raw JSON** mode.
Copy and paste the following:

```json {class="line-numbers linkable-line-numbers"}
"name" : "<your-wheeledodometry-movement-sensor-name>",
"type" : "movement_sensor",
"model" : "wheeledodometry",
"attributes" : {
    "base" : "<your-base-name>",
    "left_motors" : ["<your-base-left-motor-name-1>", "<your-base-left-motor-name-2>"],
    "right_motors" : ["<your-base-right-motor-name-1", "your-base-right-motor-name-2>"],
    "time-interval-msec": <number>
}
```

Fill in and edit the attributes as applicable.

## Attributes

The following attributes are available for `wheeledodometry` movement sensors:

| Name | Type | Inclusion | Description |
| ---- | ---- | --------- | ----------- |
| `base` | string | **Required** | The `name` of the [base](/components/base/) to which the encoders making up this movement sensor are wired. |
| `left_motors` | object | **Required** | A struct holding the name of each of the bases' left [encoded motors](/components/encoder/). |

> | `right_motors` | object | **Required** | A struct holding the name of each of the bases' right [encoded motors](/components/encoder/). |

Reviewer: Maybe use position reporting instead of encoded here.

Author: Sounds good!

| `time_interval_msec` | number | Optional | The time in between each wheeled odometry calculation. <br> Default: `500.0` </br> |

> With a configured `wheeledodometry` movement sensor, after every time `time_interval_msec` elapses during a [session](/program/apis/sessions/), your robot calculates an estimation of the position, orientation, linear velocity, and angular velocity of the wheeled base.

Author: ```suggestion
With a configured `wheeled-odometry` movement sensor, after every time `time_interval_msec` elapses during a [session](/program/apis/sessions/), your robot calculates an estimation of the position, orientation, linear velocity, and angular velocity of the wheeled base.
```

You can access these readings through the [movement sensor API](/components/movement-sensor/#api).

> After configuring a `wheeledodometry` movement sensor, you can operate your base with Viam's built-in services like the [navigation service](/services/navigation/).

Reviewer: ```suggestion
After configuring a `wheeled-odometry` movement sensor, you can operate your base with Viam's built-in services like the [navigation service](/services/navigation/).
```

Author: Looks good, done-- TYSM @  inline suggestions

> After configuring a `wheeledodometry` movement sensor, you can operate your base with Viam's built-in services like the [navigation service](/services/navigation/).

Author: ```suggestion
After configuring a `wheeled-odometry` movement sensor, you can operate your base with Viam's built-in services like the [navigation service](/services/navigation/).
```

---

> title: "Configure a wheeledodometry movement sensor"

Reviewer: ```suggestion
title: "Configure a wheeled-odometry movement sensor"
```

linkTitle: "wheeled-odometry"
type: "docs"
description: "Configure a wheeled-odometry movement sensor."
images: ["/icons/components/imu.svg"]
tags: ["movement sensor", "components", "movement sensor"]
# SMEs: Rand, Martha
---

Configure a `wheeled-odometry` movement sensor to implement _wheeled odometry_ on your robot.

_Wheeled odometry_ is the estimation of position, orientation, linear velocity, and angular velocity using the dimensions of a base.

> This model uses [encoders](/components/encoder/) from [position reporting motors](/components/motor/gpio/) to get an odometry estimate from a wheeled base.

Reviewer: ```suggestion
This model uses [encoders](/components/encoder/) from [position reporting motors](/components/motor/) to get an odometry estimate from a wheeled base.
```

## Set-up requirements

> To prepare your robot, attach [encoders](/components/encoder/) to each of the position-reporting motors on your base to measure their rotation.

Reviewer: This instruction is specific to the gpio motor model only - I would say "pick out a motor that can report it's position through encoders"

Author: Put this on line 25-- lmk if you want me to move to be in that first sentence! I also moved a lot of the info 
i had on the bottom of the page before so that people scanning can see the info about why it's useful more easily-- lmk what you think

Author: happy to add a few more sentences like `this is why this is useful` if you want!

- Configure your rover as a [wheeled base component](/components/base/wheeled/).

> - Configure each of these position-reporting motors [as `gpio` motor components](/components/motor/gpio/).

Reviewer: `Position Reporting` refers to a motor property, any motor model can have this property. The gpio motor with an encoder is one such example of a position-reporting motor. The roboclaw is a position reporting motor with encoders, the drive is always position-reporting, but a gpio motor without an encoder is not. The wheeled odometry model depends on a motor that can report its own position, usually a motor with an encoder of some sort.

Does that make a bit more sense?

Author: That makes sense! I will try and update to reflect, will comment here when done / when it's ready for re-review

Author: @randhid Ok tried to update with that advice!

"name" : "<your-wheeledodometry-movement-sensor-name>",
"type" : "movement_sensor",
"model" : "wheeledodometry",
"attributes" : {
    "base" : "<your-base-name>",
    "left_motors" : ["<your-base-left-motor-name-1>", "<your-base-left-motor-name-2>"],
    "right_motors" : ["<your-base-right-motor-name-1", "your-base-right-motor-name-2>"],
    "time-interval-msec": <number>
}
```

Fill in and edit the attributes as applicable.

## Attributes

The following attributes are available for `wheeledodometry` movement sensors:

| Name | Type | Inclusion | Description |
| ---- | ---- | --------- | ----------- |
| `base` | string | **Required** | The `name` of the [base](/components/base/) to which the encoders making up this movement sensor are wired. |
| `left_motors` | object | **Required** | A struct holding the name of each of the bases' left [position-reporting motors](/components/motor/gpio/). |
| `right_motors` | object | **Required** | A struct holding the name of each of the bases' right [position-reporting motors](/components/motor/gpio/). |
| `time_interval_msec` | number | Optional | The time in between each wheeled odometry calculation. <br> Default: `500.0` </br> |

> With a configured `wheeledodometry` movement sensor, after every time `time_interval_msec` elapses during a [session](/program/apis/sessions/), your robot calculates an estimation of the position, orientation, linear velocity, and angular velocity of the wheeled base.

Reviewer: ```suggestion
With a configured `wheeled-odometry` movement sensor, after every time `time_interval_msec` elapses during a [session](/program/apis/sessions/), your robot calculates an estimation of the position, orientation, linear velocity, and angular velocity of the wheeled base.
```

tags: ["movement sensor", "components", "movement sensor"]
# SMEs: Rand, Martha
---

Configure a `wheeled-odometry` movement sensor to implement _wheeled odometry_ on your robot.

_Wheeled odometry_ is the estimation of position, orientation, linear velocity, and angular velocity using the dimensions of a base.
This model uses [encoders](/components/encoder/) from [position reporting motors](/components/motor/) to get an odometry estimate from a wheeled base.

With a configured `wheeled-odometry` movement sensor, after every time `time_interval_msec` elapses during a [session](/program/apis/sessions/), your robot calculates an estimation of the position, orientation, linear velocity, and angular velocity of the wheeled base.
You can access these readings through the [movement sensor API](/components/movement-sensor/#api).

After configuring a `wheeled-odometry` movement sensor, you can operate your base with Viam's built-in services like the [navigation service](/services/navigation/).

## Set-up requirements

To prepare your robot, attach [encoders](/components/encoder/) to each of the position-reporting motors on your base to measure their rotation.

> - Pick out motors that can report their own position, like the [`roboclaw`](/components/motor/roboclaw/) or [`gpio` motors](/components/motor/gpio/) with [encoders](/components/encoder/#configuration). 

Author: ```suggestion
- Pick out motors that can report their own position, like the [`roboclaw`](/components/motor/roboclaw/) or [`gpio` motors](/components/motor/gpio/) with [encoders](/components/encoder/#configuration).
```
